### ISSUE TITLE
Set-Cookie header is ignored in response from url: (oAuth Callback URL).

### ISSUE THREAD

**Issue by B4TT3RY** [mitchelvanbever/remix-auth-supabase#15]
When I log in using oAuth (Twitch) I get that error in the console.

`Set-Cookie header is ignored in response from url: http://localhost:3001/login/callback?_data=routes%2Flogin.callback. The combined size of the name and value must be less than or equal to 4096 characters.`

Set-Cookie header:
```
sb=(4332 chars); Path=/; HttpOnly; SameSite=Lax
```

I used the same code as in the example. can solve this?

**Comment by rphlmr** [comment]
Hello @B4TT3RY 👋

The reason is [cookie size limit of 4kb](https://remix.run/docs/en/v1/api/remix#createcookiesessionstorage).

I test Twitch auth and see that their session token is huge (especially their `sessions.user.identities`).

Our package expects to set a Supabase [Session](https://github.com/supabase/gotrue-js/blob/6c91068bd2d5c5e4d40695ce92a5faf7a233a0ef/src/lib/types.ts#L32) in the cookie, so we can't filter per provider what to remove to be lower than 4kb.

But, because OAuth token is outside of the lib, you can remove useless data given by Twitch before sending them to your `action`.

Here an example  : 

```js
export default function OAuth() {
  const submit = useSubmit()

  useEffect(() => {
    supabaseClient.auth.onAuthStateChange((event, session) => {
      if (event === 'SIGNED_IN' && session) {
        delete session.user?.identities

        const formData = new FormData()
        formData.append('session', JSON.stringify(session))

        submit(formData, { method: 'post' })
        localStorage.removeItem('supabase.auth.token')
      }
    },
    )
  }, [submit])

  return null
}
```

I hope it helps 😇

**Comment by B4TT3RY** [comment]
The cookie has been successfully set.
Not relevant to this issue, but `const session = await oAuthStrategy.checkSession(request)` returns `null`.
Do you know any way to solve it?

```tsx
export let loader: LoaderFunction = async ({ request }) => {
  const session = await oAuthStrategy.checkSession(request)
  console.log(session)
  return {
    session,
    email: session?.user?.email,
  }
}
```
```
null
GET / 200 - - 6.569 ms
```

**Comment by rphlmr** [comment]
Your session cookie is on "Storage > Cookies" ;) 

You got null because refresh token fail (I face the same issue, setting JWT Expiry to 5s in supabase console).

If you set a `failureRedirect` in checkSession you'll be redirected. 
Without options we give you the choice to decide what to do if session is null (user no more authenticated) or is valid.

I found why ... `supabaseClient.auth.onAuthStateChange` never die and always apply after redirect.
I'm looking what could we do.
Even with this, Supabase reborn and auto refresh the token :O  (and it's conflicting with our lib because it cancel our cookie refresh token).

I'm looking at it :D

**Comment by rphlmr** [comment]
If you know how to locally test this package, could you tell me if this pull request (https://github.com/mitchelvanbever/remix-auth-supabase/pull/16) fix the issue ?

**Comment by B4TT3RY** [comment]
There seems to be a problem with UTF-8 handling.
As a result of performing base64 decode of `sb (cookie)`, only Korean has been changed to a strange language.

![image](https://user-images.githubusercontent.com/12368101/153247587-d49abc82-bc78-4c6e-aeaf-b1d4c16b30da.png)

Any korean text: `안녕하세요`

**Comment by rphlmr** [comment]
Oh I see. It fails with non "latin" characters :'( 
Can you give me a `.user_metadata.slug` example ? (any word in Korean ;) )

I want to test where it fails and maybe search on github if someone faces the same issue.

If I could be sure it's supabase issue, you could open an issue on supabase repo

**Comment by B4TT3RY** [comment]
`.user_metadata.slug` and `.user_metadata.username` are the same.
Korean is displayed normally in supabase, but there seems to be a problem in remix's CookieSessionStorage.

**Comment by rphlmr** [comment]
https://github.com/remix-run/remix/issues/1282 

I guess ?

**Comment by B4TT3RY** [comment]
Yes! And pr is https://github.com/remix-run/remix/pull/1290

**Comment by rphlmr** [comment]
Thank you to point this ! And thank you for reporting the cookie size issue :D

**Comment by B4TT3RY** [comment]
I applied #1290 PR in a patch way and tested it.
Cookies are over 4096 characters, even with deleting unnecessary data.
It seems that identifiable data should be stored in cookies and the user's profile should be stored in supabase's database.
😢

**Comment by rphlmr** [comment]
Oh sad news.

In my PR I take care to remove session.user.identities for every request (this part is huge).
Do you see other huge props on your token that should be removed ? (Maybe a screenshot blurring on sensitive data ?)

I wish I could reproduce. (My Twitch auth is working).
Maybe something in the patch makes things bigger than expected ?

I can't take the decision to store only access token and refresh token 😔 I guess some of us use user's props (email, name, email verified etc)

**Comment by rphlmr** [comment]
If you can, try to copy past your cookie value in https://mothereff.in/byte-counter
I try to fake my session with `안녕하세요😇` but can't hit the max size :/ 

Try decode it in https://www.base64decode.org/

You should not see `identities` if my fix works in your test

**Comment by B4TT3RY** [comment]
I made some changes to the code, so i need to rewrite it.
I'll write it quickly and test it out. thank you.

**Comment by B4TT3RY** [comment]
![image](https://user-images.githubusercontent.com/12368101/153541884-31382e04-bfc4-44f8-83a0-ae540ccbdc55.png)

![image](https://user-images.githubusercontent.com/12368101/153541637-87629200-e61f-4ec3-94f5-5ec0a0f93982.png)
![image](https://user-images.githubusercontent.com/12368101/153541652-271e1bf1-8ae2-47e7-b73d-16e106d8aecc.png)

Korean is not displayed normally. (It is displayed with strange characters, as in the screenshot.)
Because of that, the cookie parsing is not done properly.

![image](https://user-images.githubusercontent.com/12368101/153541992-f553b157-7ede-4682-9c00-6ba8c25784a9.png)
![image](https://user-images.githubusercontent.com/12368101/153542008-e6bf0cf9-6ed2-4efe-8da1-1128316039e7.png)
![image](https://user-images.githubusercontent.com/12368101/153542080-99b7ba12-1f91-4858-b62c-1b9725e4bb51.png)

So, when the PR of `remix-run/remix#1290` was applied and tested, the cookie size was too large.

**Comment by rphlmr** [comment]
Thanks @B4TT3RY 

Is the json below Set-Cookie the cookie value returned by remix-auth-supabase ?

If yes, you should not see "identities" props with the fix in my incoming pull request. (Unit test asserts that but who know 😅) 

Did you success following https://github.com/mitchelvanbever/remix-auth-supabase/issues/15#issuecomment-1033916889 ?
In your project node_modules/remix-auth-supabase, if npm link works, you should see the compiled code with the fix.

I know testing like that is not easy, I have to take some time to find how to publish "beta packages" 😔

**Comment by B4TT3RY** [comment]
As a result of base64 decoding `http://localhost:3001/oauth/callback?_data=routes%2Foauth.callback`'s `Set-Cookie`, `identities` still exist.

![image](https://user-images.githubusercontent.com/12368101/153551220-e589ff96-3821-4d57-a826-17e6917b8d03.png)

**Comment by B4TT3RY** [comment]
When i delete the identities and follow the PR of the remix, i get 6288 characters.

![image](https://user-images.githubusercontent.com/12368101/153552654-53695bbb-2e30-46d6-9e45-336af24edb02.png)

![image](https://user-images.githubusercontent.com/12368101/153552664-bf75df1f-38ea-4e59-872e-5785a7ca5672.png)
![image](https://user-images.githubusercontent.com/12368101/153552681-7c356b20-abc1-4e66-84a3-dbd807a6e851.png)

This works fine, but in Set-Cookie it doesn't work because it's over 4096 characters.

**Comment by rphlmr** [comment]
I'll take some time today to test all of that with my Twitch session.
6k char is really huge amount of data. (Binary conversion issue ? Will do like you and see)
Without identities, we still have many props in Session but I thought it was ok 😭.

If I can't find a solution, I have idea to remove all except token required data (and give developper a way to get user data outside of cookie), but I need more time 😇

**Comment by B4TT3RY** [comment]
Using `toBinary()` converts all strings to `\x00` format. So the cookie seems to be increasing.

**Comment by mwarnerdotme** [comment]
> Ok I got the point.
> 
> I have to update sample and their readme.
> 
> First, you have to set additional options when creating supabaseClient to prevent auto refresh by the sdk :
> 
> ```js
> export const supabaseClient = createClient(
>   window.env.SUPABASE_URL,
>   window.env.PUBLIC_SUPABASE_ANON_KEY,
>  { autoRefreshToken: false, persistSession: false },
> )
> ```
> 
> After that it'll work until a token refresh occur. BUT, when we refresh the token, in our lib, we get back a full bloated token and ... guess what ... cookie size > 4kb 🤪
> 
> I'll make a pull request. I think this issue should be reopen to track ;)
> 
> @mitchelvanbever we get back huge token from supabase :/ Maybe we can "purge" useless data in Session, like `session.user?.identities`

Just wanted to say that you helped me a lot by pointing on the client options object. I think the `autoRefreshToken` in particular is a setting that's intended for most browser use cases (when set to `true`, which is the default), but our primary use case is not in the browser when we're working with Remix. I kept seeing Auth requests in my Supabase overview dashboard without intentionally authenticating, and I realize now that it was automatically getting new refresh tokens on the server-side without me knowing, which counted as a hit of course. Imagine if I had tried to deploy that at scale - yikes!

This was also causing an issue where I could not use refresh tokens to sign in (ex: after an access_token had expired, I'm using the `jsonwebtokens` library to verify signatures and expired tokens from Supabase on every protected route request) because my refresh tokens were invalidated every time I tried to use them (because of the autoRefreshToken setting).

Googler's from the future, hopefully I helped you because I just spent a few hours scratching my head on this.

**Comment by rphlmr** [comment]
Should be fixed soon by https://github.com/mitchelvanbever/remix-auth-supabase/pull/28

**Comment by mitchelvanbever** [comment]
Reduced the cookie size by only storing `user.id` instead of the entire user, fetching the user should be a separate call.

Released in V4.0.0
